### ISSUE TITLE
Docker provider: vagrant halt should also stop the docker host machine if there are no more running containers

### ISSUE THREAD

**Issue by cweagans** [mitchellh/vagrant#6832]
When I run `vagrant halt`, I expect my environment to stop running. However, if I'm using the Docker provider on a Mac, the Docker host VM is still running in the background and there's nothing that indicates that.

i propose that if I run `vagrant halt`, the following should happen:

* All containers are stopped as normal
* If there are no more containers running on the docker host VM, the docker host VM should also be stopped.
* If there are more containers running on the docker host VM (i.e. if it's shared across projects or something), leave it running, but tell the user about it.

Ideally, I'd also be able to individually control the host VM as I would in any other multi-machine setup, but that's a separate issue, I suppose.

In addition, sharing a VM across projects is kind of a non-requirement in my mind. No other vagrant setup will let you do that to my knowledge, so I'd also be okay with just indiscriminately shutting down the docker host VM even if there are other containers running on it, but I suspect that somebody somewhere is using this functionality now, so if it's not possible to remove, then meh.

**Comment by sethvargo** [comment]
Hi @cweagans 

Thank you for opening an issue. The automatic installation and booting of the boot2docker VM for working with Vagrant and Docker is a fairly hard requirement at this time. I understand that your particular use case does not concern itself whether the VM remains running, but it was an often-requested feature that Docker "just works" out of the box. The boot2docker instance uses very few resources and makes for a very pleasurable development experience for the 95% use case.

If you'd like a lot of control over the actual boot2docker instance, I would suggest just booting that VM yourself with public networking and just export the DOCKER_HOST environment variable to run your docker commands directly. Does that make sense? Thanks! :smile:

**Comment by cweagans** [comment]
Okay, but "just works" implies that vagrant should do the same thing in most circumstances. If I run vagrant halt, my environment should halt and stop using resources. Also, we don't use boot2docker. We use coreos for the docker host machine, and while that's still a relatively low-resource VM, I still don't want it running in the background all the time.

If I open a PR to stop the docket host machine after vagrant halt finishes (possibly with an environment variable to opt out of that), would it be accepted? I've got most of the code working as a plugin already and there's not a lot to it.

**Comment by sethvargo** [comment]
Hi @cweagans 

Please understand that, with the Docker provider, Vagrant intentionally masks a lot of magic. For most users, we don't even want them to feel like there exists a Virtual Machine. Instead, we want the experience to feel as though the user is running Docker natively on their host operating system.

For power users who want full control over the virtual machine that is actually running Docker, it is much easier to just create that machine as a VM and manage it's lifecycle like you would any other machine.

If you already have a plugin, that's great. We always love to see plugins :smile:. If you want to open source the plugin, that'd be great too!

**Comment by cweagans** [comment]
@sethvargo I think we're talking past each other here. I don't think it's reasonable for vagrant to turn on a VM and not tell me about it when I run `vagrant up`, and then proceed to not turn off the VM when I run `vagrant halt`. That is not what people expect, nor should it be. Vagrant should use the resources it needs to use to bring up the environment specified in the Vagrantfile, and then *stop* using them when it's told to. "It doesn't use that many resources" doesn't mitigate that.

I also don't think this should be a separate plugin that people have to download. It should be the default behavior.

Basically, I'm just asking that vagrant clean up after itself and I don't understand why there's pushback on that, especially given that I'm willing to implement it.

Sorry if I sound hostile - that's definitely not my intent. I'm just a little confused as to why this issue was closed when it's still a problem.

**Comment by cweagans** [comment]
And to further clarify - I'm not asking that the boot2docker functionality be removed or anything like that. I think it's great that Vagrant makes this seamless and easy. I just want it to clean up after it's done - the b2d VM shouldn't *continue* running after I stop my environment.
